### PR TITLE
Raft heartbeat process in event base

### PR DESCRIPTION
**Pull request by critical27**
Our workflow looks like below, as for an write request

| leader bgWorker | leader ioThreadPool | leader threadMananger | follower ioThreadPool | follower threadManager |
| --------------- | ------------------- | --------------------- | --------------------- | ---------------------- |
|                 | 1. requestReceived     |                       |                       |                        |
|                 |                     | 2. processInThread       |                       |                        |
| (heartbeat)     | 3. replicate           |                       |                       |                        |
|                 |                     |                       | 4. receive appendLog       |                        |
|                 |                     |                       |                       | 5. process      |
|                 | 6. collectN            |                       |                       |                        |
|                 |                     | 7. commit                |                       |                        |

From the view of a storaged process, there would be many partition, we could be blocked in **any** phase if it takes longer time to process. Phase 5 and 7 are the main cause of blocking as before. When pressure is big enough, it is possible that all worker thread is busy. This would raise a lot problems, one of the most notorious among them is leader change. 

Block reason:
1. lock
2. rocksdb write stall (while holding lock)

IMO, there should be only one phase could be blocked, which is leader commit (phase 7). To achieve that, many works need to do:
1. The second commit `don not hold raft lock when commit` will relax the restriction of raft lock. (we have another lock replicatingLogs to prevent concurrent replicate)
2. The third commit `follower delay commit if write stall` will make follower commit logs not to block (phase 5, follower could commit in async)
3. The first commit `heartbeat refactor` make heartbeat process in event base, even if all worker threads are blocked, there won't be unexpected election.

So, the heartbeat logic is:
1. If a log is on the fly, only send a heartbeat which will only be processed in eventbase.
2. If no log is on the fly, both a dummy log(previous behavior) and a heartbeat will be sent.

depends on https://github.com/vesoft-inc/nebula-common/pull/497

### CODE REVIEW

File: src/kvstore/raftex/RaftPart.cpp [hunk 1]
```diff
@@ -1680,36 +1648,46 @@ void RaftPart::processAppendLogRequest(
...
-    if (req.get_committed_log_id() > committedLogId_) {
+    LogID lastLogIdCanCommit = std::min(lastLogId_, req.get_committed_log_id());
+    if (lastLogIdCanCommit > committedLogId_) {
         // Commit some logs
         // We can only commit logs from firstId to min(lastLogId_, leader's commit log id),
         // follower can't always commit to leader's commit id because of lack of log
-        LogID lastLogIdCanCommit = std::min(lastLogId_, req.get_committed_log_id());
-        CHECK_LE(committedLogId_ + 1, lastLogIdCanCommit);
-        if (commitLogs(wal_->iterator(committedLogId_ + 1, lastLogIdCanCommit))) {
+        auto code = commitLogs(wal_->iterator(committedLogId_ + 1, lastLogIdCanCommit), false);
```
Comment (bright-starry-sky) [inline src/kvstore/raftex/RaftPart.cpp:1656]: One question, when` wait == false` , we will ignore whether the log commits successfully or not.but why do you need to set the committedLogId_ at line 1671?

File: src/kvstore/raftex/Host.cpp [hunk 2]
```diff
@@ -494,18 +478,83 @@ folly::Future<cpp2::AppendLogResponse> Host::sendAppendLogRequest(
...
+            if (t.hasException()) {
+                cpp2::HeartbeatResponse resp;
+                resp.set_error_code(cpp2::ErrorCode::E_EXCEPTION);
+                pro.setValue(std::move(resp));
+                return;
+            } else {
+                pro.setValue(std::move(t.value()));
+            }
+        });
+    return promise.getFuture();
```
Comment (liuyu85cn) [inline src/kvstore/raftex/Host.cpp:525]: may be moved(not sure about this)?

File: src/kvstore/raftex/RaftPart.cpp (hunk 1 above)
Comment (critical27) [reply src/kvstore/raftex/RaftPart.cpp:1656]: L1671 only set committedLogId_ in resp, but we didn't update the `committedLogId_`, when we commit successfully, we update the committedLogId_, and response the new one (Line 1661 1662).

**Review by bright-starry-sky — APPROVED** [review]

**Review by liuyu85cn — APPROVED** [review]

File: src/kvstore/raftex/Host.cpp (hunk 2 above)
Comment (critical27) [reply src/kvstore/raftex/Host.cpp]: fixed

**Pull request merged**